File: Learn_Basic/src/learn_cmake/src/service_opencv/opencv_find_point_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "learn_interface/srv/topic_opencv.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.hpp"

#include <opencv2/opencv.hpp>
#include <chrono>
#include <thread>

class SubscriberNode : public rclcpp::Node
{
public:
    SubscriberNode(const std::string &name)
        : Node(name)
    {
        client_ = this->create_client<learn_interface::srv::TopicOpencv>(
            "learn_service_opencv_find_point");
    }

    bool wait_for_server(int timeout_sec = 60)
    {
        RCLCPP_INFO(this->get_logger(), "等待服务可用...");
        if (client_->wait_for_service(std::chrono::seconds(timeout_sec)))
        {
            RCLCPP_INFO(this->get_logger(), "服务连接成功");
            return true;
        }
        else
        {
            RCLCPP_ERROR(this->get_logger(), "服务在 %d 秒内不可用", timeout_sec);
            return false;
        }
    }

    rclcpp::Client<learn_interface::srv::TopicOpencv>::SharedFuture send_request(bool scan_once = true)
    {
        if (!client_->service_is_ready())
        {
            RCLCPP_ERROR(this->get_logger(), "服务未就绪");
            return rclcpp::Client<learn_interface::srv::TopicOpencv>::SharedFuture();
        }

        auto request = std::make_shared<learn_interface::srv::TopicOpencv::Request>();
        request->scan_once = scan_once;

        auto future_and_id = client_->async_send_request(request);
        return future_and_id.future.share();
    }

    void handle_msg(const std::shared_ptr<learn_interface::srv::TopicOpencv::Response> response)
    {
        if (!response)
        {
            RCLCPP_ERROR(this->get_logger(), "无响应数据");
            return;
        }

        // 处理图像
        try
        {
            cv::Mat cv_image = cv_bridge::toCvCopy(response->image, "bgr8")->image;
            cv::imshow("Received Image", cv_image);
            cv::waitKey(1);
        }
        catch (const cv_bridge::Exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "图像错误: %s", e.what());
        }

        // 处理点
        if (!response->points.empty())
        {
            RCLCPP_INFO(this->get_logger(), "检测到 %zu 个点", response->points.size());
            for (size_t i = 0; i < response->points.size(); ++i)
            {
                const auto &point = response->points[i];
                RCLCPP_INFO(this->get_logger(),
                            "点%zu: (%.1f, %.1f, %.1f)",
                            i + 1, point.x, point.y, point.z);
            }
        }
    }

private:
    rclcpp::Client<learn_interface::srv::TopicOpencv>::SharedPtr client_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto client = std::make_shared<SubscriberNode>("opencv_find_point_service_node");

    if (!client->wait_for_server())
    {
        return 0;
    }

    int fps = 24;

    while (rclcpp::ok())
    {
        auto future = client->send_request(true);
        if (future.valid())
        {
            if (rclcpp::spin_until_future_complete(client, future) ==
                rclcpp::FutureReturnCode::SUCCESS)
            {
                auto response = future.get();
                client->handle_msg(response);
            }
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(1000 / fps));
    }

    cv::destroyAllWindows();
    rclcpp::shutdown();
    return 0;
}

File: Learn_Basic/src/learn_py/learn_py/service_opencv/opencv_find_point_client.py
import cv2
import rclpy
from rclpy.node import Node
from learn_interface.srv import TopicOpencv
from cv_bridge import CvBridge
import time

class SubscriberNode(Node):
    def __init__(self, name):
        super().__init__(name)
        self.client = self.create_client(TopicOpencv, "learn_service_opencv_find_point")
        self.bridge = CvBridge()
        
    def wait_for_server(self, timeout_sec = 60):
        self.get_logger().info('等待服务可用...')
        if self.client.wait_for_service(timeout_sec=timeout_sec):
            self.get_logger().info('服务连接成功')
            return True
        else:
            self.get_logger().error(f'服务在 {timeout_sec} 秒内不可用')
            return False
        
    
    def send_request(self, scan_once=True):
        if not self.client.service_is_ready():
            self.get_logger().error('服务未就绪')
            return None
            
        request = TopicOpencv.Request()
        request.scan_once = scan_once
        
        # self.get_logger().info(f'发送请求: scan_once={scan_once}')
        
        future = self.client.call_async(request)
        return future
        
    def handle_msg(self, response):
        
        if not response:
            self.get_logger().error('无响应数据')
            return False
        
        
        # 处理图像
        try:
            cv_image = self.bridge.imgmsg_to_cv2(response.image, "bgr8")
            cv2.imshow("Received Image", cv_image)
            cv2.waitKey(1)
        except Exception as e:
            self.get_logger().error(f"图像错误: {str(e)}")
        
        # 处理点
        if response.points:
            self.get_logger().info(f"检测到 {len(response.points)} 个点")
            for i, point in enumerate(response.points):
                self.get_logger().info(f"点{i+1}: ({point.x:.1f}, {point.y:.1f} {point.z:.1f})")

def main(args=None):
    rclpy.init(args=args)
    client = SubscriberNode("opencv_find_point_service_node")
    
    if not client.wait_for_server():
        return
    
    fps = 24
    
    while rclpy.ok():
        future = client.send_request(scan_once=True)
        if future:
            rclpy.spin_until_future_complete(client, future)
            response = future.result()
            client.handle_msg(response)
        
        time.sleep(1 / fps)
    
    try:
        rclpy.spin(client)
    except KeyboardInterrupt:
        pass
    finally:
        cv2.destroyAllWindows()
        client.destroy_node()
        rclpy.shutdown()
